rpcs3/evdev_joystick_handler.cpp

```diff
@@ -13,560 +17,755 @@
-                    }
-                }
-                else
-                    pad->m_sticks[axis].m_value = scale_axis(evt.code, value);
-            }
-            break;
-        }
-        default:
-            LOG_ERROR(GENERAL, "Unknown joystick #%d event %d", i, evt.type);
-            break;
-        }
-    }
+	update_devs();
+
+	for (auto& device : devices)
+	{
+		auto pad = device.pad;
+		auto axis_orientations = device.axis_orientations;
+		auto& dev = device.device;
+		if (dev == nullptr) continue;
+
+		// Try to query the latest event from the joystick.
+		input_event evt;
+		int ret = libevdev_next_event(dev, LIBEVDEV_READ_FLAG_NORMAL, &evt);
+
+		// Grab any pending sync event.
+		if (ret == LIBEVDEV_READ_STATUS_SYNC)
+		{
+			LOG_NOTICE(GENERAL, "Captured sync event");
+			ret = libevdev_next_event(dev, LIBEVDEV_READ_FLAG_NORMAL | LIBEVDEV_READ_FLAG_SYNC, &evt);
+		}
+
+		if (ret < 0)
+		{
+			// -EAGAIN signifies no available events, not an actual *error*.
+			if (ret != -EAGAIN)
+				LOG_ERROR(GENERAL, "Failed to read latest event from joystick: %s [errno %d]", strerror(-ret), -ret);
+			continue;
+		}
+
+		bool is_negative = false;
+		int value;
+		int button_code = GetButtonInfo(evt, dev, value, is_negative);
+		if (button_code < 0 || value < 0)
+			continue;
+
+		bool is_button_or_trigger = evt.type == EV_KEY || button_code == ABS_Z || button_code == ABS_RZ;
+
+		// Translate any corresponding keycodes to our normal DS3 buttons and triggers
+		for (int i = 0; i < static_cast<int>(pad->m_buttons.size()); i++)
+		{
+			if (pad->m_buttons[i].m_keyCode != button_code)
+				continue;
+
+			// Be careful to handle mapped axis specially
+			if (evt.type == EV_ABS)
+			{
+				// get axis direction and skip on error or set to 0 if the stick/hat is actually pointing to the other direction.
+				// maybe mimic on error, needs investigation. FindAxisDirection should ideally never return -1 anyway
+				int direction = FindAxisDirection(axis_orientations, i);
+				if (direction < 0)
+				{
+					LOG_ERROR(HLE, "FindAxisDirection = %d, Button Nr.%d, value = %d", direction, i, value);
+					continue;
+				}
+				else if (direction != (is_negative ? 1 : 0))
+				{
+					pad->m_buttons[i].m_value = 0;
+					pad->m_buttons[i].m_pressed = 0;
+					continue;
+				}
+			}
+
+			pad->m_buttons[i].m_value = static_cast<u16>(value);
+			TranslateButtonPress(button_code, pad->m_buttons[i].m_pressed, pad->m_buttons[i].m_value);
+		}
+
+		// Translate any corresponding keycodes to our two sticks. (ignoring thresholds for now)
+		for (int idx = 0; idx < static_cast<int>(pad->m_sticks.size()); idx++)
+		{
+			bool pressed_min = false, pressed_max = false;
+			u16 val_min = 0, val_max = 0;
+
+			// m_keyCodeMin is the mapped key for left or down
+			if (pad->m_sticks[idx].m_keyCodeMin == button_code)
+			{
+				bool is_direction_min = false;
```

Comment:
With this:
Left Stick Left: LY+
Left Stick Down: LX-
Left Stick Right: LY-
Left Stick Up: LX+
My device.stick_val[idx] is always 0 on the up-down axis (x-y axis goes from -255 to +255)
With this:
Left Stick Left: LY-
Left Stick Down: LX-
Left Stick Right: LY+
Left Stick Up: LX+
My device.stick_val[idx] moves from -255 to +255 on the up-down axis (x-y axis ok).

Reply:
should be okay now

Reply:
ok. With this change, I always get a signal now. It's still a bit buggy (I have to manually edit the config file to have the correct movements and, with my custom stick, when I move to the right, it registers 1st a small move to the right, then a big move). Now, nothing is a stopper for me anymore. So, if you want to merge, that's fine with me. We can improve this with investigation and subsequent PR. In the present state, it's certainly already a nice improvement to the existing situation.

Reply:
try again. I fixed the lag in the dialog